openforcefield/topology/molecule.py

```diff
@@ -1542,6 +1542,38 @@ def __init__(self,
                     other)
                 raise Exception(msg)
 
+        # Validate the molecule
+        self._validate()
+
+    @property
+    def has_unique_atom_names(self):
+        """True if the molecule has unique atom names, False otherwise."""
+        unique_atom_names = set([atom.name for atom in self.atoms])
+        if len(unique_atom_names) < self.n_atoms:
+            return False
+        return True
+
+    def generate_unique_atom_names(self):
+        """
+        Generate unique atom names using element name and number of times that element has occurred
+        e.g. 'C1', 'H1', 'O1', 'C2', ...
+
+        """
+        from collections import defaultdict
+        element_counts = defaultdict(int)
+        for atom in self.atoms:
+            symbol = atom.element.symbol
+            element_counts[symbol] += 1
+            atom.name = symbol + str(element_counts[symbol])
+
+    def _validate(self):
+        """
+        Validate the molecule, ensuring it has unique atom names
+
+        """
+        if not self.has_unique_atom_names:
+            self.generate_unique_atom_names()
```

Comment:
I like the idea of having a general place to put molecule-validation tasks. I'm also in favor of providing functionality to add unique atom names. But _always_ generating unique atom names (in some cases overwriting ones that are already there), with no way to disable the behavior, may be going too far.

Reply:
Once we ditch ParmEd, it will no longer be necessary to generate unique atom names. But until we replace it, we're stick with the need to do so if we want our parameters to be deliverable to different molecular simulation packages.

That means that this needs to either happen to _all_ molecules we load into the toolkit, or it means we need to create more code to manage the transformation to other packages that does this silently behind the scenes. That may be an alternative that is workable here!